tests/SAML2/XML/samlp/StatusCodeTest.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SAML2\XML\samlp;
+
+use SAML2\Constants;
+use SAML2\DOMDocumentFactory;
+use SAML2\Utils;
+
+/**
+ * Class \SAML2\XML\samlp\StatusCodeTest
+ *
+ * @author Tim van Dijen, <tvdijen@gmail.com>
+ * @package simplesamlphp/saml2
+ */
+class StatusCodeTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @return void
+     */
+    public function testMarshalling(): void
+    {
+        $subCode = new StatusCode();
+        $subCode->setValue(Constants::STATUS_REQUEST_DENIED);
+
+        $statusCode = new StatusCode();
+        $statusCode->setValue(Constants::STATUS_RESPONDER);
+        $statusCode->setSubCodes([$subCode]);
+
+        $document = DOMDocumentFactory::fromString('<root />');
+        /** @psalm-var \DOMElement $document->firstChild */
+        $statusCodeElement = $statusCode->toXML($document->firstChild);
+
+        $this->assertEquals(Constants::STATUS_RESPONDER, $statusCodeElement->getAttribute('Value'));
+
+        /** @psalm-var \DOMElement[] $statusCodeElements */
+        $statusCodeElements = Utils::xpQuery($statusCodeElement, './saml_protocol:StatusCode');
+        $this->assertEquals(Constants::STATUS_REQUEST_DENIED, $statusCodeElements[0]->getAttribute('Value'));
```

Comment:
One idea that I was considering is to make all these classes inherit from a basic class that allows getting a string with the XML. Something like an interface with an abstract `toXML()`, and an implementation of `__toString()` similar to this: 

```php
    public function __toString()
    {
        $xml = $this->toXML();
        return $xml->ownerDocument->saveXML($xml);
    }
```

That would ease tests, because we then just need to compare the resulting XML with an actual string, eliminating the need to process XML manually ourselves.

Reply:
Yeah that would be great! We can add an abstract fromXML to the interface as well